modules/apps/29-fee/ibc_module.go

```diff
@@ -47,14 +47,24 @@ func (im IBCModule) OnChanOpenInit(
 	}
 
 	if versionMetadata.FeeVersion != types.Version {
-		return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s, got %s", types.Version, versionMetadata.FeeVersion)
+		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "expected %s, got %s", types.Version, versionMetadata.FeeVersion)
+	}
+
+	appVersion, err := im.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID, chanCap, counterparty, versionMetadata.AppVersion)
+	if err != nil {
+		return "", err
+	}
+
+	versionMetadata.AppVersion = appVersion
+	versionBytes, err := types.ModuleCdc.MarshalJSON(&versionMetadata)
+	if err != nil {
+		return "", err
 	}
 
 	im.keeper.SetFeeEnabled(ctx, portID, channelID)
 
 	// call underlying app's OnChanOpenInit callback with the appVersion
-	return im.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID,
-		chanCap, counterparty, versionMetadata.AppVersion)
+	return string(versionBytes), nil
```

Comment:
Why was the call to underlying app's `OnChanOpenInit` removed here?

Reply:
Its moved to above on line 53 so that the returned application version (for example: `ics20-1`) can be inserted into the `AppVersion` field of `types.Metadata`, and re-marshalled to a JSON encoded version string. Otherwise we would return directly from the app callback (writing `ics20-1` as the channel version into state), we would expect ics29 to be enabled but subsequent handshake callbacks on would fail as the channel version in state would not be the metadata object:

```json
{
    "AppVersion": "ics20-1",
    "FeeVersion": "ics29-1",
}
```

Does that make sense?